README.md
# Algorithm::Statistic
This is Perl module Algorithm::Statistic. Provides the following functions:

* k order statistic

* median

INSTALLATION

Type the following command:

    $ curl -L http://cpanmin.us | perl - Algorithm::Statistic

Or install cpanm and then run the following command to install
Algorithm::Statistic:

    $ cpanm Algorithm::Statistic

If you get an archive of this distribution, unpack it and build it
as per the usual:

    $ tar xzf Algorithm::Statistic-$version.tar.gz
    $ cd Algorithm::Statistic-$version
    $ perl Makefile.PL
    $ make && make test

Then install it:

    $ make install

DOCUMENTATION

Algorithm::Statistic documentation is available as in POD. So you can do:

    $ perldoc Algorithm::Statistic

to read the documentation online with your favorite pager.

TODO
* Add MULTICALL for user specified compare CV's.

LICENSE AND COPYRIGHT

Copyright (c) 2015, Igor Karbachinsky. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
0.03
- redundant copying from AV* to std::vector removed.
- compare sub argument moved to the end of args list and has default implementation for numeric SV*.
- g++ in Makefile movede to c++ for generalization.
- warns changed on croaks
- O2 optimizations enabled for c++.


lib/Algo.pm
package Algo;

use strict;
use warnings;
use XSLoader;

use Exporter 5.57 'import';

our $VERSION     = '0.01';
our %EXPORT_TAGS = ( 'all' => [qw<kth_order_statistic mediana>] );
our @EXPORT_OK   = ( @{ $EXPORT_TAGS{'all'} } );

XSLoader::load('Algo', $VERSION);

1;

=head1 NAME

Algo - algorithms library

=head1 VERSION

Version 0.01

=head1 SYNOPSIS

    use Algo;

=head1 DESCRIPTION

This module provides several math and statistical algorithms implementations in C++ XS for perl.

=head1 Functions

=head2 C<< kth_order_statistic(compare_sub, array_ref, k) >>

This function allows to find k-th order statistic for certain array of elements. 
Note that this function changes input array (like std::nth_element im STL C++) according to the next rule:
element at k-th position will become k-th order atatistic. Each element from the left of k will be less then k-th
and each from the right will be greater. This algorithm works with linear complexity O(n).

    $statistic = kth_order_statistic(\&compare, $array_ref, $k);

For example C<compare> function could be simple comparison for integers:

    sub compare {
        $_[0] <=> $_[1]
    }

=head2 C<< mediana(compare_sub, array_ref) >>

This function allows to find mediana for certain array of elements. This method is the same as n/2 kth order statistc.
Like C<kth_order_statistic> this functions changes input array according to the same rule.

    $mediana = mediana(\&compare, $array_ref);

=head1 BUGS

If you find a bug please contact me via email.

=head1 AUTHOR

Igor KarbachinskyE <igorkarbachinsky@mail.ru>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2015, Igor Karbachinsky. All rights reserved.

=cut


include/kth_order_statistic.hpp
#ifndef ALGO_KTH_ORDER_STATISTIC_HPP
#define ALGO_KTH_ORDER_STATISTIC_HPP

/* K-th order statistic evaluator
 * 
 * The same as nth_element in stl library.
 * 
 * Author: I. Karbachinsky <igorkarbachinsky@mail.ru> 2014
 */


#include <iostream>
#include <cassert>
#include <stdexcept>
#include <iterator>
#include <functional>

#include "debug.hpp"

namespace algo {

template<class InputIterator, class Compare=std::minus<typename InputIterator::value_type> >
const InputIterator KthOrderStatistic(const InputIterator begin, const InputIterator end, size_t k, const Compare &compare) {
    typedef typename std::iterator_traits<InputIterator>::value_type T;

    size_t n = std::distance(begin, end);

    DEBUG("", std::endl);
    DEBUG_ITERABLE("Array", begin, end);
    DEBUG("k", k)
    DEBUG("n", n)

    if (n == 0)
        throw std::invalid_argument("Empty array passed");

    if ( k > n)
        throw std::invalid_argument("Bad argument k. Maybe it's out of range");

    auto it = begin,
         jt = end-1;

    auto pivot = begin + n/2;
    const T pivot_value = *pivot; 

    DEBUG("Pivot value", pivot_value);

    do {
        while (compare(*it, pivot_value) < 0) {
            ++it;
        }
        while (compare(*jt, pivot_value) > 0) {
            --jt;
        }

        if (it < jt) {
            DEBUG("Swap it", *it);
            DEBUG("Swap jt", *jt);

            iter_swap(it, jt);

            DEBUG_ITERABLE("Array after rearranging", begin, end);

            if (compare(*it, *jt) == 0 && jt - it == 1)
                break;
        }
    } while (it < jt);

    pivot = jt;

    size_t pivot_n = std::distance(begin, pivot);

    DEBUG("Pivot position after rearranging", pivot_n);

    if (pivot_n == k) {
        DEBUG("FOUND on positon: ", pivot_n);
        return pivot;
    }
    if (pivot_n > k) 
        return KthOrderStatistic<InputIterator, Compare>(begin, pivot, k, compare);
    
    return KthOrderStatistic<InputIterator, Compare>(pivot, end, k - pivot_n, compare);
}

} // namespace algo

#endif


lib/Algorithm/Statistic.pm
package Algorithm::Statistic;

use strict;
use warnings;
use XSLoader;

use Exporter 5.57 'import';

our $VERSION     = '0.03';
our %EXPORT_TAGS = ( 'all' => [qw<kth_order_statistic median>] );
our @EXPORT_OK   = ( @{ $EXPORT_TAGS{'all'} } );

XSLoader::load('Algorithm::Statistic', $VERSION);

1;

=head1 NAME

Algorithm::Statistic - different statistical algorithms library

=head1 VERSION

Version 0.03

=head1 SYNOPSIS

    use Algorithm::Statistic qw/:all/;

=head1 DESCRIPTION

This module provides several math and statistical algorithms implementations in C++ XS for perl.

=head1 Functions

=head2 C<< kth_order_statistic(array_ref, k [, \&compare]) >>

This function allows to find k-th order statistic for certain array of elements. 
Note that this function changes input array (like std::nth_element im STL C++) according to the next rule:
element at k-th position will become k-th order atatistic. Each element from the left of k will be less then k-th
and each from the right will be greater. This algorithm works with linear complexity O(n). By default you don't have to specify comparator
for integers and float numbers.

    my $statistic = kth_order_statistic($array_ref, $k);

But in more complex cases it's posible to specify comparator.

    my $statistic_cmp = kth_order_statistic($array_ref, $k, \&compare);

For example C<compare> function could be simple comparison for strings:

    sub compare {
        $_[0] cmp $_[1]
    }

=head2 C<< median(array_ref, \&compare) >>

This function allows to find median for certain array of elements. This method is the same as n/2 kth order statistc.
Like C<kth_order_statistic> this function changes input array according to the same rule.

    my $median = median($array_ref [, \&compare]);

=head1 BUGS

If you find a bug please contact me via email.

=head1 AUTHOR

Igor Karbachinsky <igorkarbachinsky@mail.ru>

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2015, Igor Karbachinsky. All rights reserved.

=cut
